haskell/test/SMT/Common/Gen.hs

```diff
@@ -0,0 +1,424 @@
+      , mkCall (XtzForbidden $ Freeze (#amount .! voterFreezeAmt) ) Nothing
+      , mkCall proposeAction (Just 1)
+      , mkCall (XtzForbidden $ Vote permitProtecteds) (Just 1)
+      , mkCall flushAction (Just 2)
+      , mkCall (XtzForbidden $ Unfreeze (#amount .! freezeAmt) ) Nothing
+      , mkCall dropProposalAction Nothing
+      ] <> ((\param -> mkCall (XtzAllowed $ CallCustom param) Nothing) <$> mkCustomCalls args)
+  where
+    getVoteFreezeAmt permitProtecteds =
+      foldl' (\i (PermitProtected voteParam _) -> (voteParam & vVoteAmount) + i) 0 permitProtecteds
+
+genVote :: GeneratorT (ProposalKey -> Natural -> [PermitProtected VoteParam])
+genVote = do
+  mkGenVoteParams <- vectorOf 5 $ genVoteParams
+  pure $ \proposalKey correctVoteAmt -> (\f -> f proposalKey correctVoteAmt) <$> mkGenVoteParams
+  where
+    genVoteParams :: GeneratorT (ProposalKey -> Natural -> PermitProtected VoteParam)
+    genVoteParams = do
+      userPool <- get <&> gsAddresses
+      sender_ <- Gen.element userPool
+
+      invalidFrom <- Gen.element userPool
+      from <- Gen.element [sender_, invalidFrom]
+
+      voteType <- Gen.bool
+
+      let mkVoteParam proposalKey voteAmt = VoteParam
+            { vProposalKey = proposalKey
+            , vVoteType = voteType
+            , vVoteAmount = voteAmt
+            , vFrom = fst from
+            }
+
+      mkSigned <- sign' (snd from)
+
+      mkMaybe <- Gen.element [\_ -> Nothing, \a -> Just a]
+
+      let mkPermit signed =
+            Permit (toPublic . snd $ from) (TSignature signed)
+
+      pure $ \proposalKey correctVoteAmt ->
+          let voteParam = mkVoteParam proposalKey correctVoteAmt
+              ds = DataToSign
+                  { dsChainId = dummyChainId
+                  , dsContract =
+                      ContractAddress $ unsafeParseContractHash "KT1WsLzQ61xtMNJHfwgCHh2RnALGgFAzeSx9"
+                  , dsNonce = Nonce 0
+                  , dsData = voteParam
+                  }
+              signed = mkSigned (lPackValueRaw ds)
+          in PermitProtected voteParam (mkMaybe $ mkPermit signed)
+
+-- | A version of `Tezos.Crypto.sign` to be run in `MonadGen`
+sign' :: SecretKey -> GeneratorT ( ByteString -> Signature)
+sign' addr = do
+  seed <- drgNewSeed . seedFromInteger <$> Gen.integral (Range.linearFrom 0 -1000 1000)
+  pure $ \bytes -> fst $ withDRG seed $ do
+    sign addr bytes
+
+genFlush :: GeneratorT Parameter
+genFlush = do
+  amt <- Gen.integral (Range.constant 0 100)
+  pure $ XtzForbidden $ Flush amt
+
+genDropProposal :: GeneratorT (ProposalKey -> Parameter)
+genDropProposal = do
+  pure $ \key -> XtzForbidden $ Drop_proposal key
+
+genUpdateDelegate :: GeneratorT Parameter
+genUpdateDelegate = do
+  params <- genDelegateParams
+  pure $ XtzForbidden $ Update_delegate params
+
+genTransferContractTokens :: GeneratorT (Address -> Parameter)
+genTransferContractTokens = do
+  let transfers = []
```

Comment:
Should this transfer some tokens sometimes?

Reply:
Nice catch, I've fixed this now